### DISCUSSION TITLE
proxy auth support

### DISCUSSION THREAD

**Discussion by cjyar** [photoprism/photoprism#3845] [General]
Does photoprism support proxy auth like what's provided by [oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy)? The idea is to hide an app like photoprism behind an https proxy, and have the proxy block unauthenticated traffic from reaching the upstream app. The proxy handles authentication, and once the user is authenticated, requests from their session are passed upstream. The app (photoprism) needs to support reading an HTTP header like `Remote-User` or `Remote-Group` and trusting it.

**Comment by graciousgrey** [comment]
oauth2 proxy is not supported at this time. However, we are currently working on OpenID Connect support.

**Comment by cjyar** [comment]
Thanks for your reply. Would you be interested in a PR to support auth proxy?

**Reply by lastzero** [reply]
Since OpenID Connect is based on OAuth, I suggest waiting until we have implemented/shipped it. Supporting OAuth without using OIDC is probably not much work then, if any.

**Reply by cjyar** [reply]
Sorry; I think my terminology was vague. I'm talking about auth proxies in general, of which oauth2-proxy is a common example. To support auth proxies, photoprism would just need to be configurable to trust a `Remote-User` or similar header. It doesn't require photoprism to understand OIDC or OAuth2; and it allows photoprism to indirectly support any authentication mechanism that the auth proxy can handle, not just OIDC.

Without looking at the photoprism code, I would assume that adding auth proxy support would be unrelated to OIDC support, and would mostly not create merge conflicts. It should be a relatively small change.

Having said that, OIDC will probably fit my use case, so I could just wait for you to implement that. But if you're interested in general auth proxy support, I could work on it.

**Reply by lastzero** [reply]
So basically trust any X-User header added by a proxy without being able to block clients (after failed authentication) or invalidate sessions (after privilege level changes)?

**Reply by cjyar** [reply]
Basically, but with some caveats:

- The auth proxy would be responsible for blocking clients after failed authentication, so photoprism wouldn't have to manage that.
- The header(s) to trust should probably be configurable; but often the auth proxy can be configured to send whatever header is desired, so configurable header name isn't a hard requirement.
- The admin certainly wouldn't want photoprism to trust any headers unless it's hiding behind an auth proxy; that would create a vulnerability like CVE-2023-22515. In other words, photoprism's code to support auth proxy headers should definitely be disabled by default.
- [Grafana](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/auth-proxy/) does a good job of configuring auth proxy headers; it allows configuring which header to trust, and what that header's contents correspond to: user or email.
- Some software will also trust a `Remote-Group` header.

I don't know the best way for auth proxy code to interact with session code. The auth proxies I've worked with allow the user to log out without notifying the upstream application. Any session cookies set by the application would be passed through, unmodified by the auth proxy. So maybe the application would invalidate the session and create a new one if the contents of the user header change?

**Reply by lastzero** [reply]
Seems similar to this?

- https://github.com/photoprism/photoprism/discussions/2117

The bottom line is that there are many different server endpoints that require different permissions and handling, so just putting a standard proxy in front of it (without a lot of testing and customization effort) probably won't work - at least if you want the same functionality as with the built-in authentication and authorization checks.

**Reply by cjyar** [reply]
Yes, very much that! Thanks for your patience with the Q&A here, and for pointing me to that discussion. That answers all of my questions.
